src/main/java/com/baidu/hugegraph/util/TaskStatusCache.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2017 HugeGraph Authors
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership. The ASF
+ * licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.baidu.hugegraph.util;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.baidu.hugegraph.api.task.TaskAPI;
+import com.baidu.hugegraph.structure.Task;
+
+public class TaskStatusCache {
+
+    private static final Task FAKE_TASK = new Task();
+
+    private Map<String, Map<Long, Task>> tasks;
+    private Map<String, AtomicBoolean> querying;
+    private ScheduledExecutorService service;
+
+    private static TaskStatusCache instance = new TaskStatusCache();
+
+    private TaskStatusCache() {
+        this.tasks = new ConcurrentHashMap<>();
+        this.querying = new HashMap<>();
+        this.service = null;
+    }
+
+    public static TaskStatusCache instance() {
+        return instance;
+    }
+
+    public void add(TaskAPI api, long taskId) {
+        Map<Long, Task> tasks = this.tasksOfGraph(api.graph());
+        if (!tasks.containsKey(taskId)) {
+            tasks.putIfAbsent(taskId, FAKE_TASK);
+        }
+        this.scheduleQueryTask(api);
+    }
+
+    public Task get(TaskAPI api, long taskId) {
+        this.add(api, taskId);
+        String graph = api.graph();
+        return this.tasks.get(graph).get(taskId);
+    }
+
+    public void remove(TaskAPI api, long taskId) {
+        Map<Long, Task> tasks = this.tasksOfGraph(api.graph());
+        tasks.remove(taskId);
+        if (tasks.isEmpty()) {
+            this.tasks.remove(api.graph());
+        }
+        if (this.tasks.isEmpty()) {
+            this.querying = new HashMap<>();
+            this.service.shutdown();
+        }
+    }
+
+    private void scheduleQueryTask(TaskAPI api) {
```

Comment:
prefer to schedule a timer and let it traversal all graphs and fetch tasks for each graph